Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,243 @@
+        }).fail(failTask);
+    } else {
+        defer.resolve(null);
+    }
+
+    return defer.promise;
+}
+
+
+async function uploadFiles() {
+    tl.debug('connected to ftp host:' + serverEndpointUrl.host);
+
+    tl.debug('files to process: ' + files.length);
+
+    var defer: Q.Deferred<void> = Q.defer<void>();
+
+    cleanRemoteIfRequired().then(() => {
+        var promises = [];
+        var nestedPromises = [];
+        promises.push(ftpHelper.createRemoteDirectory(remotePath).then(() => {
+            directoriesProcessed++;
+            printStatus('remote directory successfully created/verified: ' + remotePath);
+        })); // ensure root remote location exists
+
+        files.forEach((file) => {
+            tl.debug('file: ' + file);
+            var remoteFile: string = preservePaths ?
+                path.join(remotePath, file.substring(rootFolder.length)) :
+                path.join(remotePath, path.basename(file));
+
+            remoteFile = remoteFile.replace(/\\/gi, "/"); // use forward slashes always
+            tl.debug('remoteFile: ' + remoteFile);
+
+            var stats = tl.stats(file);
+            if (stats.isDirectory()) { // create directories if necessary
+                promises.push(ftpHelper.createRemoteDirectory(remoteFile).then(() => {
+                    directoriesProcessed++;
+                    printStatus('remote directory successfully created/verified: ' + remoteFile);
+                }));
+            } else if (stats.isFile()) { // upload files
+                if (overwrite) {
+                    promises.push(ftpHelper.uploadFile(file, remoteFile).then(() => {
+                        filesUploaded++;
+                        printStatus('successfully uploaded: ' + file + ' to: ' + remoteFile);
+                    }));
+                } else {
+                    promises.push(ftpHelper.remoteExists(remoteFile).then((exists: boolean) => {
+                        if (!exists) {
+                            nestedPromises.push(ftpHelper.uploadFile(file, remoteFile).then(() => {
+                                filesUploaded++;
+                                printStatus('successfully uploaded: ' + file + ' to: ' + remoteFile);
+                            }));
+                        } else {
+                            filesSkipped++;
+                            printStatus('skipping file: ' + file + ' remote: ' + remoteFile + ' because it already exists');
+                        }
+                    }));
+                }
+            }
+        });
+
+        Q.all(promises).then(() => {
+            Q.all(nestedPromises).then(() => {
+                tl.setResult(tl.TaskResult.Succeeded, 'FTP upload successful' + getFinalStatusMessage());
+                ftpClient.end();
+                ftpClient.destroy();
+                defer.resolve(null);
+            }).fail(failTask);
+        }).fail(failTask);
+    }).fail(failTask);
+}
+
+ftpClient.on('ready', uploadFiles);
```

Comment:
verified; new version is pushed up.  This is definitely clearer this way.  Not sure if further improvement can be made the way promises are handled in uploadFiles(), though I think the error handling now correctly propagates up to doWork() for all cases and is handled correctly there.